sql: use the correct locking strength for FOR SHARE clause

arulajmani opened the pull request:
Previously, FOR SHARE and FOR KEY SHARE would use non-locking KV scans.
Now that the lock table supports shared locks, we can use lock.Shared as
the locking strength for KV scans. This patch does that, and in doing
so, wires up SHARED locks end to end.

By default, we turn of this functionality for serializable transactions.
Instead, it's gated behind a session setting called
`enable_shared_locking_for_serializable`.

Closes https://github.com/cockroachdb/cockroach/issues/91545

Release note (sql change): SELECT FOR SHARE and SELECT FOR UPDATE
previously did not acquire any locks. Users issuing these statements
would expect them to acquire shared locks (multiple readers allowed,
no writers though). This patch switches over the behavior to acquire
such read locks.

For serializable transactions, we default to the old behaviour, unless
the `enable_shared_locking_for_serializable` session setting is set
to true. We'll probably switch this behavior in the future, but for
now, given shared locks are a preview feature, we gate things behind
a session setting.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/109638)
<!-- Reviewable:end -->

arulajmani commented:
@michae2 it looks like the `optbuilder` and `execbuilder` tests already handle the `FOR SHARE` clause correctly. Are there any other tests we would want to add here? 

Separately, @nvanbenschoten and @michae2, do we want to add a cluster setting to revert back to the pre-23.2`FOR SHARE` behavior?

nvb reviewed (commented):
> Separately, [@nvanbenschoten](https://github.com/nvanbenschoten) and [@michae2](https://github.com/michae2), do we want to add a cluster setting to revert back to the pre-23.2`FOR SHARE` behavior?

That sounds like a good idea.

> Reviewed 10 of 10 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-:-NczHIIF8BBEsj10m661:b-v1qf1r)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @michae2)

___
*[`pkg/sql/row/locking.go` line 30 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NczGdgj94WNgQUZU6aY:-NczGdgj94WNgQUZU6aZ:b-ad46ev) ([raw file](https://github.com/cockroachdb/cockroach/blob/438fdf6878cdc0d883c89df6d07c1adef78513dc/pkg/sql/row/locking.go#L30)):*
> ```Go
> 		fallthrough
> 	case descpb.ScanLockingStrength_FOR_SHARE:
> 		return lock.Shared
> ```

Do we need to hide this behind a cluster version check?
___
*[`pkg/sql/logictest/testdata/logic_test/shared_locks` line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NczGnk63qgLujv4xqIX:-NczGnk63qgLujv4xqIY:b-9x2wm) ([raw file](https://github.com/cockroachdb/cockroach/blob/438fdf6878cdc0d883c89df6d07c1adef78513dc/pkg/sql/logictest/testdata/logic_test/shared_locks#L1)):*
> ```
> statement ok
> ```

Should we call this file `select_for_share` to mirror the `select_for_update` file?


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
I'm going to let this run through CI to catch any plumbing issues, but should be good for a look. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-:-NfSTa4G8kCAr59aQ06b:bo3d1zd)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @nvanbenschoten)

___
*[`pkg/sql/row/locking.go` line 30 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NczGdgj94WNgQUZU6aY:-NfRyW-Q9xRDdSZU435m:b-onp83l) ([raw file](https://github.com/cockroachdb/cockroach/blob/438fdf6878cdc0d883c89df6d07c1adef78513dc/pkg/sql/row/locking.go#L30)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we need to hide this behind a cluster version check?
</blockquote></details>

Added a cluster version check and hid this behaviour behind a cluster setting (defaults to 23.1 behaviour).
___
*[`pkg/sql/logictest/testdata/logic_test/shared_locks` line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NczGnk63qgLujv4xqIX:-NfRydbU3uhZZTjQwQQ7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/438fdf6878cdc0d883c89df6d07c1adef78513dc/pkg/sql/logictest/testdata/logic_test/shared_locks#L1)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we call this file `select_for_share` to mirror the `select_for_update` file?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

michae2 reviewed (changes requested):
I should have commented on this earlier, I'm sorry I didn't. `GetKeyLockingStrength` is too low to make the decision about whether to use locking or not. By this point we should already have picked correctly based on session variables, version gates, isolation level, etc, and `GetKeyLockingStrength` should merely be translating from execinfrapb locking to kv locking. These decisions about whether to lock live in optbuilder. I think the place to put the check would be in `buildScan`: https://github.com/cockroachdb/cockroach/blob/af05d816a1b0b060947114b22d400862b89383c5/pkg/sql/opt/optbuilder/select.go#L700-L724

After `locking.get()` we know the locking strength the optimizer wants. This would be a good spot to silently turn off locking if `FOR SHARE` has been requested and we're not using RC and we can't do it due to a version gate.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-:-NfSX6oV1PaWJrsh7O-B:b5pfaia)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`pkg/sql/row/locking.go` line 34 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NfSZHBv0aA-R2BLSzIK:-NfSZHBv0aA-R2BLSzIL:b-i509gf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5fc49c029977fec0a153912e91e728dfe8b1458/pkg/sql/row/locking.go#L34)):*
> ```Go
> 	false,
> 	settings.WithPublic,
> )
> ```

Whenever the optimizer makes decisions based on settings, (a) they should be session settings rather than cluster settings, and (b) they need to be added to `Memo.IsStale` in `pkg/sql/opt/memo/memo.go`. There's an example of that in this PR: https://github.com/cockroachdb/cockroach/pull/105857/commits/bb857aae80e1f20f67b457e7a8e790f208aa8cb5


<!-- Sent from Reviewable.io -->

arulajmani committed:
sql: use the correct locking strength for FOR SHARE clause

Previously, FOR SHARE and FOR KEY SHARE would use non-locking KV scans.
Now that the lock table supports shared locks, we can use lock.Shared as
the locking strength for KV scans. This patch does that, and in doing
so, wires up SHARED locks end to end.

By default, we turn off this functionality for serializable transactions.
Instead, it's gated behind a session setting called
`enable_shared_locking_for_serializable`.

Informs #91545

Release note (sql change): SELECT FOR SHARE and SELECT FOR KEY SHARE
previously did not acquire any locks. Users issuing these statements
would expect them to acquire shared locks (multiple readers allowed,
no writers though). This patch switches over the behavior to acquire
such read locks.

For serializable transactions, we default to the old behaviour, unless
the `enable_shared_locking_for_serializable` session setting is set
to true. We'll probably switch this behavior in the future, but for
now, given shared locks are a preview feature, we gate things behind
a session setting.

arulajmani reviewed (commented):
Thanks for the pointers @michae2. This should be good for a look now. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-:-NfTOw-V71B8aq6II3-n:b-iuwrs1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @nvanbenschoten)

___
*[`pkg/sql/row/locking.go` line 34 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NfSZHBv0aA-R2BLSzIK:-NfTP20N4XxZ2RCNH3WZ:bgap9jx) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5fc49c029977fec0a153912e91e728dfe8b1458/pkg/sql/row/locking.go#L34)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Whenever the optimizer makes decisions based on settings, (a) they should be session settings rather than cluster settings, and (b) they need to be added to `Memo.IsStale` in `pkg/sql/opt/memo/memo.go`. There's an example of that in this PR: https://github.com/cockroachdb/cockroach/pull/105857/commits/bb857aae80e1f20f67b457e7a8e790f208aa8cb5
</blockquote></details>

Makes sense. I've added a session variable, similar in spirit to the one you added for durable locks for serializable transactions.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
Looks like there are some failures in CI. Where possible, I think we should just set the session variable to avoid the test churn.

The failures that look like the following are more interesting:
> expected "cannot execute FOR SHARE in a read-only transaction", but no error occurred

I guess the change to `Builder.buildScan` is masking the check in `Builder.buildLocking`. I don't think we want SELECT FOR SHARE to be usable in read-only transactions, even if the session variable is unset. If we let it succeed when the session variable is unset then we risk breaking apps if we change the default (or remove the variable entirely) in the future.

Maybe we should call `buildLocking` as a form of error checking before discarding the locking mode in `buildScan`. What do you think @michae2?

> Reviewed 14 of 33 files at r3, 29 of 29 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-:-NfTXt2n-7L9ZZ7M_eS_:b-bcxb4k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @michae2)

___
*[`-- commits` line 9 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NfTVNhs1ORCnl4flt9q):*
s/of/off/
___
*[`-- commits` line 15 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NfTVRzfBpm8fEqKgp1g):*
Heh, you'll want to help the docs writers out a little more than that. Maybe also mention the session variable and how this is still a preview feature.
___
*[`pkg/sql/BUILD.bazel` line 582 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NfTVtZC5KspgF_gHhle:-NfTVtZC5KspgF_gHhlf:br8u5de) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f16a0696dc6f5c72477c0bc6a992746c095769d/pkg/sql/BUILD.bazel#L582)):*
> ```Starlark
>         "//pkg/util/uint128",
>         "//pkg/util/uuid",
>         "@com_github_cockroachdb_apd//:apd",
> ```

Not sure why this got pulled in again.
___
*[`pkg/sql/opt/optbuilder/select.go` line 720 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NfTWKiA2fBsBIQvM9AG:-NfTWKiA2fBsBIQvM9AH:bdi1zrl) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f16a0696dc6f5c72477c0bc6a992746c095769d/pkg/sql/opt/optbuilder/select.go#L720)):*
> ```Go
> 			private.Locking.Durability = tree.LockDurabilityGuaranteed
> 		}
> 		// Check if we can actually use shared locks here , or we need to use
> ```

`s/here ,/here,/`
___
*[`pkg/sql/sessiondatapb/local_only_session_data.proto` line 426 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NfTWksWBcgLdQVuDpYg:-NfTWksWBcgLdQVuDpYh:b-idmyw2) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f16a0696dc6f5c72477c0bc6a992746c095769d/pkg/sql/sessiondatapb/local_only_session_data.proto#L426)):*
> ```proto
>   // better performance.) Weaker isolation levels always use durable locking.
>   bool durable_locking_for_serializable = 109;
>   // SharedLockingForSerializable, if set to true, means SELECT FOR  SHARE and
> ```

nit: double space between "FOR" and "SHARE".
___
*[`pkg/sql/sessiondatapb/local_only_session_data.proto` line 428 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NfTWrZ83TumFwvm_Whe:-NfTWrZ83TumFwvm_Whf:b-73oty3) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f16a0696dc6f5c72477c0bc6a992746c095769d/pkg/sql/sessiondatapb/local_only_session_data.proto#L428)):*
> ```proto
>   // SharedLockingForSerializable, if set to true, means SELECT FOR  SHARE and
>   // SELECT FOR KEY SHARE statements issued by transactions that run with
>   // Serializable isolation will acquiring shared locks; otherwise, they'll
> ```

nit: lower-case Serializable.
___
*[`pkg/sql/sessiondatapb/local_only_session_data.proto` line 428 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NfTWykEBDRIsNhbNkuM:-NfTWykEBDRIsNhbNkuN:bl4mk6) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f16a0696dc6f5c72477c0bc6a992746c095769d/pkg/sql/sessiondatapb/local_only_session_data.proto#L428)):*
> ```proto
>   // SharedLockingForSerializable, if set to true, means SELECT FOR  SHARE and
>   // SELECT FOR KEY SHARE statements issued by transactions that run with
>   // Serializable isolation will acquiring shared locks; otherwise, they'll
> ```

s/acquiring/acquire/
___
*[`pkg/sql/opt/exec/execbuilder/testdata/select_for_update` line 45 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NfTXkl-9t_dj8NX_tfl:-NfTXkl-9t_dj8NX_tfm:bpd5iyx) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f16a0696dc6f5c72477c0bc6a992746c095769d/pkg/sql/opt/exec/execbuilder/testdata/select_for_update#L45)):*
> ```
>   locking strength: for no key update
> 
> # By default, SELECT FOR SHARE doesn't acquire any locks.
> ```

It's nice that this is now reflected in the EXPLAIN output!
___
*[`pkg/sql/logictest/testdata/logic_test/select_for_share` line 47 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NfTXNYm3yX5b1Pxci78:-NfTXNYm3yX5b1Pxci79:bs0bmyw) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f16a0696dc6f5c72477c0bc6a992746c095769d/pkg/sql/logictest/testdata/logic_test/select_for_share#L47)):*
> ```
> # TODO(arul): Until https://github.com/cockroachdb/cockroach/issues/107766 is
> # addressed, we'll incorrectly report shared locks as having "Exclusive" lock
> # strength; however, having this query in here is useful to make sure there are
> ```

We'll also incorrectly only report a single "granted" lock holder, right? That's why we only see two rows instead of three?
___
*[`pkg/sql/logictest/testdata/logic_test/select_for_share` line 112 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NfTXZ1CCmnKYsnHKxX7:-NfTXZ1CCmnKYsnHKxX8:b-8t1xcv) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f16a0696dc6f5c72477c0bc6a992746c095769d/pkg/sql/logictest/testdata/logic_test/select_for_share#L112)):*
> ```
> 
> # TODO(arul): Add a test to show that the session setting doesn't apply to read
> # committed transactions. We currently can't issue SELECT FOR SHARE statmenets
> ```

"statements"


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
> The failures that look like the following are more interesting:

> expected "cannot execute FOR SHARE in a read-only transaction", but no error occurred


I was looking into this as well -- what's interesting is that just setting the session variable doesn't get rid of it, which Im surprised by. Changing the global default does. My best guess is that this is because I've added this session to "local only session data", instead of session data. 

This may be a side quest, but I'll see if making the switch over fixes things. This might be interesting, as it relates to the `DurableLockingForSerializable` field. I could also be doing something stupid here.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-:-NfTbZcu8OofYMPedF5N:b-bova73)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @nvanbenschoten)

___
*[`pkg/sql/BUILD.bazel` line 582 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NfTVtZC5KspgF_gHhle:-NfTaOe904QuQBPiKTL7:b-80u85a) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f16a0696dc6f5c72477c0bc6a992746c095769d/pkg/sql/BUILD.bazel#L582)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Not sure why this got pulled in again.
</blockquote></details>

My GoLand keeps doing this for some reason. It keeps replacing `apd/v3` with `apd`, so every time I re-run `./dev generate` this happens. I'll make sure things are clean before I actually get to bors-ing this.
___
*[`pkg/sql/logictest/testdata/logic_test/select_for_share` line 47 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NfTXNYm3yX5b1Pxci78:-NfTbQ3H6xhych_3Dzrb:b-aml4ft) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f16a0696dc6f5c72477c0bc6a992746c095769d/pkg/sql/logictest/testdata/logic_test/select_for_share#L47)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We'll also incorrectly only report a single "granted" lock holder, right? That's why we only see two rows instead of three?
</blockquote></details>

Ah right, I missed that -- didn't realize we include waiters in here as well. I'll update this comment.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> they need to be added to `Memo.IsStale` in `pkg/sql/opt/memo/memo.go`

Maybe this is related? There's no update to `memo.go` in this PR.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-:-NfTdYpi7zHHKbWzWaMa:blje0u7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @michae2)



<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
> Maybe this is related? There's no update to `memo.go` in this PR.

Yup, you're right. I'll fix this in my next push.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-:-NfTf2HK-ZrtcDhZZiA3:bzsse9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @nvanbenschoten)



<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
Alright, addressed all the comments here and worked through some of the CI flakes. I think the main ones left now are related to:

```
I guess the change to `Builder.buildScan` is masking the check in `Builder.buildLocking`. I don't think we want SELECT FOR SHARE to be usable in read-only transactions, even if the session variable is unset. If we let it succeed when the session variable is unset then we risk breaking apps if we change the default (or remove the variable entirely) in the future.
```

I agree. @michae2, I'll defer to you on calling `buildLocking` for error checking or rolling out something a bit more special cased here.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-:-NfTk_SE0YKRF4UDtD9D:b-sdonk9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @nvanbenschoten)

___
*[`-- commits` line 15 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NfTVRzfBpm8fEqKgp1g):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Heh, you'll want to help the docs writers out a little more than that. Maybe also mention the session variable and how this is still a preview feature.
</blockquote></details>

Haha changed this to actually be a useful release note. Good call on the preview call out.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/select_for_update` line 45 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NfTXkl-9t_dj8NX_tfl:-NfTjJmfBdGkBepYab8i:ban2it0) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f16a0696dc6f5c72477c0bc6a992746c095769d/pkg/sql/opt/exec/execbuilder/testdata/select_for_update#L45)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

It's nice that this is now reflected in the EXPLAIN output!
</blockquote></details>

Heh I never considered the old thing to be a bug until you pointed it out!


<!-- Sent from Reviewable.io -->

michae2 reviewed (changes requested):
Thank you for turning this around so quickly! I've got fixes for the remaining CI failures here, if you want them: https://github.com/cockroachdb/cockroach/pull/111478

> I agree. [@michae2](https://github.com/michae2), I'll defer to you on calling `buildLocking` for error checking or rolling out something a bit more special cased here.

I think we just need to add a small check to `optbuilder.(*Builder).validateLockingInFrom`. I've done that in the draft PR I linked to, as well.

> Reviewed 7 of 29 files at r5, 7 of 10 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-:-NfUWkfw2EY9MsNgTyYZ:b-fw5kkw)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`-- commits` line 15 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NfTVRzfBpm8fEqKgp1g):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Haha changed this to actually be a useful release note. Good call on the preview call out.
</blockquote></details>

I think "SELECT FOR UPDATE" should be removed from this sentence. (It previously did acquire locks, right?)
___
*[`pkg/sql/opt/optbuilder/select.go` line 701 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NfUK5hn1ejrzkEsf_AX:-NfUK5hn1ejrzkEsf_AY:b-jyw7uq) ([raw file](https://github.com/cockroachdb/cockroach/blob/eeb93c89831d8f933c70f2ac74369a1c7b1dbdd8/pkg/sql/opt/optbuilder/select.go#L701)):*
> ```Go
> 	if locking.isSet() {
> 		private.Locking = locking.get()
> 		// TODO(arul): This needs a cluster version check.
> ```

Gah, good catch! I should have done this. If you want to, you can put me as the TODO.
___
*[`pkg/sql/opt/optbuilder/select.go` line 725 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NfUHq3d88KWrtR7oT-h:-NfUHq3d88KWrtR7oT-i:b-6rrimz) ([raw file](https://github.com/cockroachdb/cockroach/blob/eeb93c89831d8f933c70f2ac74369a1c7b1dbdd8/pkg/sql/opt/optbuilder/select.go#L725)):*
> ```Go
> 			// Shared locks weren't a thing prior to v23.2, so we must use non-locking
> 			// reads.
> 			if !b.evalCtx.Settings.Version.IsActive(context.TODO(), clusterversion.V23_2) ||
> ```

`b.ctx` instead of `context.TODO()`
___
*[`pkg/sql/opt/optbuilder/select.go` line 736 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NfUIxO8B_lAD0E68iGd:-NfUIxO8B_lAD0E68iGe:b-7bba8w) ([raw file](https://github.com/cockroachdb/cockroach/blob/eeb93c89831d8f933c70f2ac74369a1c7b1dbdd8/pkg/sql/opt/optbuilder/select.go#L736)):*
> ```Go
> 				// won't be too happy if we start asking for nonsensical "replicated
> 				// non-locking locks".
> 				private.Locking.Durability = tree.LockDurabilityBestEffort
> ```

I think we should also set `private.Locking.WaitPolicy = tree.LockWaitBlock` as well. And should also clear `private.Locking.Form`.

(In fact, it might be more straightforward to simply do `private.Locking = opt.Locking{}` instead of setting individual fields.)
___
*[`pkg/sql/row/locking.go` line 34 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NfSZHBv0aA-R2BLSzIK:-NfUKLq3-LodJnXwa2qO:bff479j) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5fc49c029977fec0a153912e91e728dfe8b1458/pkg/sql/row/locking.go#L34)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Makes sense. I've added a session variable, similar in spirit to the one you added for durable locks for serializable transactions.
</blockquote></details>

Thank you for doing this!
___
*[`pkg/sql/row/locking.go` line 37 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NfUKsJN3Xhcq8LmIEt0:-NfUKsJN3Xhcq8LmIEt1:b-rof8h) ([raw file](https://github.com/cockroachdb/cockroach/blob/eeb93c89831d8f933c70f2ac74369a1c7b1dbdd8/pkg/sql/row/locking.go#L37)):*
> ```Go
> 	case descpb.ScanLockingStrength_FOR_UPDATE:
> 		// We currently perform exclusive per-key locking when FOR_UPDATE is
> 		// used because Update locks have not yet been implemented.
> ```

suggestion: I think we should remove this old comment, too.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
> I've got fixes for the remaining CI failures here, if you want them: [![](https://avatars.githubusercontent.com/u/32038551?s=60&v=4) Sl sql integration #111478](https://github.com/cockroachdb/cockroach/pull/111478)

Thanks a lot Michael, this was breezy! I cherry picked your changes and merged them into my commit. The only test that's failing now is the one with prepared statements -- it makes sense, because by the time we come around to executing the statement in a transaction (which is the point at which we know it's read-only), we've already thrown away the locking information. 

One option here would be to store an additional boolean on `opts.Locking` struct -- maybe a `opts.Locking.OriginalLockingStrength`. We could then use this for our read-only transaction rejection check. There may be something cleaner than this, though, so @michae2 I'll defer to you on the approach here.

> Dismissed @michae2 from a discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-:-NfWbcJl13nqv1KA6K19:b-3bwf5y)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @nvanbenschoten)

___
*[`-- commits` line 15 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NfTVRzfBpm8fEqKgp1g):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I think "SELECT FOR UPDATE" should be removed from this sentence. (It previously did acquire locks, right?)
</blockquote></details>

I'd meant to say "SELECT FOR KEY SHARE". Fixed.
___
*[`pkg/sql/opt/optbuilder/select.go` line 701 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NfUK5hn1ejrzkEsf_AX:-NfWWF8b8JCtZUc-8h3W:b-tebxrp) ([raw file](https://github.com/cockroachdb/cockroach/blob/eeb93c89831d8f933c70f2ac74369a1c7b1dbdd8/pkg/sql/opt/optbuilder/select.go#L701)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Gah, good catch! I should have done this. If you want to, you can put me as the TODO.
</blockquote></details>

I'll try and follow this up with a quick patch, but just in case I don't get to it before I head out for OOO, I'll add both ours.
___
*[`pkg/sql/opt/optbuilder/select.go` line 725 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NfUHq3d88KWrtR7oT-h:-NfWWUZ234ckw3V3EMMC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/eeb93c89831d8f933c70f2ac74369a1c7b1dbdd8/pkg/sql/opt/optbuilder/select.go#L725)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

`b.ctx` instead of `context.TODO()`
</blockquote></details>

Done.
___
*[`pkg/sql/opt/optbuilder/select.go` line 736 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NfUIxO8B_lAD0E68iGd:-NfWWzq53IUAZh9dKJ4u:b-pftvy8) ([raw file](https://github.com/cockroachdb/cockroach/blob/eeb93c89831d8f933c70f2ac74369a1c7b1dbdd8/pkg/sql/opt/optbuilder/select.go#L736)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I think we should also set `private.Locking.WaitPolicy = tree.LockWaitBlock` as well. And should also clear `private.Locking.Form`.

(In fact, it might be more straightforward to simply do `private.Locking = opt.Locking{}` instead of setting individual fields.)
</blockquote></details>

Good call; switched to `private.Locking = opt.Locking{}`
___
*[`pkg/sql/row/locking.go` line 34 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NfSZHBv0aA-R2BLSzIK:-NfWX1i_ERFXGSAOWiLu:b-t76dif) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5fc49c029977fec0a153912e91e728dfe8b1458/pkg/sql/row/locking.go#L34)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Thank you for doing this!
</blockquote></details>

Thanks for the pointers! This approach is much nicer.
___
*[`pkg/sql/row/locking.go` line 37 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NfUKsJN3Xhcq8LmIEt0:-NfWXPj5CPxHthcolPGT:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/eeb93c89831d8f933c70f2ac74369a1c7b1dbdd8/pkg/sql/row/locking.go#L37)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

suggestion: I think we should remove this old comment, too.
</blockquote></details>

Removed.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 6 of 10 files at r6, 8 of 8 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109638)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @michae2)

___
*[`pkg/sql/opt/optbuilder/select.go` line 724 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NfWj2SkBXhNus-B3Eof:-NfWj2SkBXhNus-B3Eog:b-m4hnqx) ([raw file](https://github.com/cockroachdb/cockroach/blob/0191d4b3693f564ca66a4ba1820af456dc0e9ef8/pkg/sql/opt/optbuilder/select.go#L724)):*
> ```Go
> 			// Shared locks weren't a thing prior to v23.2, so we must use non-locking
> 			// reads.
> 			//if !b.evalCtx.Settings.Version.IsActive(b.ctx, clusterversion.V23_2) ||
> ```

Why did this get commented out? Intentional?


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109638)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @nvanbenschoten)

___
*[`pkg/sql/opt/optbuilder/select.go` line 724 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NfWj2SkBXhNus-B3Eof:-NfWnDFcEBEcACNdKkKE:b7lsntp) ([raw file](https://github.com/cockroachdb/cockroach/blob/0191d4b3693f564ca66a4ba1820af456dc0e9ef8/pkg/sql/opt/optbuilder/select.go#L724)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why did this get commented out? Intentional?
</blockquote></details>

No, I was trying to see if this branch was tripping up a test failure and forgot to uncomment the thing. :facepalm: 


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
Ah, drat, prepared statements. :thinking: There's also the possibility that we would prepare with a different isolation level or different `enable_shared_locking_for_serializable` setting than when we execute the prepared statement.

Let's try moving your code from `optbuilder.(*Builder).buildScan` to [this part](https://github.com/cockroachdb/cockroach/blob/d0c8f31cf40a4682d3f59ad37fed1fc48ace4102/pkg/sql/opt/exec/execbuilder/relational.go#L2924) of `execbuilder.(*Builder).buildLocking`. That should only run when executing the prepared statement.

Then we should also be able to get rid of the small read-only check I added to `optbuilder.(*Builder).validateLockingInFrom`.

Now that you've pointed this out, I think I will also have to move my `DurableLockingForSerializable` check there as well, but I want to hold off on that until after https://github.com/cockroachdb/cockroach/pull/110945 goes in (or perhaps as part of it) because I need to think about how it will interact with implicit locking in UPDATE etc.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-:-NfWpsGe50GHysbNV-mc:bn35hgl)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)



<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
(Also I think you'll have to do a `git restore -s HEAD^ pkg/sql/opt/optbuilder/testdata/*` after moving the code from optbuilder to execbuilder.)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-:-NfWwjcsAKkO7KkCqn9n:b-f3txs)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)



<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
Yup, that worked out quite well. I'll let this run through CI one more time to make sure I didn't break anything (I tested a few tests locally, and things seemed fine). Should be good for another look. 

> Now that you've pointed this out, I think I will also have to move my `DurableLockingForSerializable` check there as well.

Good point! 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-:-NfXBUYL7Ew-QXCGhU7r:b-1yytt9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @nvanbenschoten)



<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 10 files at r6, 3 of 8 files at r7, 4 of 16 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-:-NfXIHPLEGx1Fw0WCqEk:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`pkg/sql/opt/exec/execbuilder/relational.go` line 642 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NfXH2UqDfrzxQvWMxch:-NfXH2UqDfrzxQvWMxci:b-xzfazu) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c7d471e2c304ece170f9f7a119a01b0e5394290/pkg/sql/opt/exec/execbuilder/relational.go#L642)):*
> ```Go
> 			locking = opt.Locking{}
> 		}
> 	}
> ```

Oh, did this not work if it was inside `buildLocking`? We'll need it to be inside, because by this point in the optimizer `SELECT FOR SHARE` could have turned into a locking join rather than a locking scan.

(I can do this part in another PR if you need to go.)


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109638)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @michae2 and @nvanbenschoten)

___
*[`pkg/sql/opt/exec/execbuilder/relational.go` line 642 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/109638#-NfXH2UqDfrzxQvWMxch:-NfXSJ406h7_mxD1LOkx:b-jklmjb) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c7d471e2c304ece170f9f7a119a01b0e5394290/pkg/sql/opt/exec/execbuilder/relational.go#L642)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Oh, did this not work if it was inside `buildLocking`? We'll need it to be inside, because by this point in the optimizer `SELECT FOR SHARE` could have turned into a locking join rather than a locking scan.

(I can do this part in another PR if you need to go.)
</blockquote></details>

This wasn't intentional, I moved it inside. The other thing I realized was that we need to be careful with setting `b.ContainsNonDefaultKeyLocking`. Not doing so was causing the test timeouts in CI. 


<!-- Sent from Reviewable.io -->

arulajmani commented:
Alright, all tests are passing. Thanks again for all the pointers @michae2.

bors r=michae2
